File: claude-usage-panel@fschmutz.github.io/lib/pure/snapshots.js
// The session snapshots `claudectl session` keeps (claude-code/tabs.js), as
// the preferences window shows them. The CLI owns the store - one JSON file
// per label, {version, savedAt, sessions: [{name, cwd, session_id}]} - and
// this only reads it, ordered the same way the CLI's `store` lists it.

/** Autosaves are the ones the 30-minute schedule writes. */
export const AUTO_SNAPSHOT_PREFIX = 'auto-';

/**
 * Summarise parsed snapshot files, newest first by savedAt, then label in
 * reverse code-point order (not localeCompare: the Swift port must agree).
 * Mirrored by ClaudeUsageCore/Snapshots.swift; tests/fixtures/snapshots.json.
 * @param {{label: string, data: any}[]} files one entry per `<label>.json`
 * @returns {{count: number, autos: number, newest: {label, savedAt, sessions}|null}}
 */
export function summarizeSnapshots(files) {
    const valid = files
        .filter(f => f.data && Array.isArray(f.data.sessions))
        .map(f => ({label: f.label, savedAt: Number(f.data.savedAt) || 0, sessions: f.data.sessions}))
        .sort((a, b) => b.savedAt - a.savedAt || (a.label === b.label ? 0 : (a.label < b.label ? 1 : -1)));
    return {
        count: valid.length,
        autos: valid.filter(s => s.label.startsWith(AUTO_SNAPSHOT_PREFIX)).length,
        newest: valid[0] ?? null,
    };
}

File: macos/Sources/ClaudeUsageCore/Snapshots.swift
import Foundation

// The session snapshots `claudectl session` keeps (claude-code/tabs.js), as the
// Settings window shows them. The CLI owns the store - one JSON file per label,
// {version, savedAt, sessions: [{name, cwd, session_id}]} - and this only
// reads it. Mirrors the GNOME extension's lib/pure/snapshots.js and is pinned
// to it by tests/fixtures/snapshots.json.

public enum Snapshots {
    /// Autosaves are the ones the 30-minute schedule writes.
    public static let autoPrefix = "auto-"

    public struct Entry: Equatable, Sendable {
        public let label: String
        /// Milliseconds since the epoch, 0 when the file has none.
        public let savedAtMs: Double
        public let names: [String]
    }

    public struct Summary: Equatable, Sendable {
        public let count: Int
        public let autos: Int
        public let newest: Entry?
    }

    /// Summarise parsed snapshot files, newest first by savedAt, then label in
    /// reverse code-point order. `json` is the parsed file, nil when unreadable.
    public static func summarize(_ files: [(label: String, json: Any?)]) -> Summary {
        let valid: [Entry] = files.compactMap { file in
            guard let obj = file.json as? [String: Any],
                let sessions = obj["sessions"] as? [[String: Any]]
            else { return nil }
            let savedAt = (obj["savedAt"] as? NSNumber)?.doubleValue ?? 0
            return Entry(
                label: file.label, savedAtMs: savedAt,
                names: sessions.map { $0["name"] as? String ?? "" })
        }
        .sorted {
            $0.savedAtMs != $1.savedAtMs
                ? $0.savedAtMs > $1.savedAtMs
                : $1.label.unicodeScalars.lexicographicallyPrecedes($0.label.unicodeScalars)
        }
        return Summary(
            count: valid.count,
            autos: valid.filter { $0.label.hasPrefix(autoPrefix) }.count,
            newest: valid.first)
    }
}
